src/Promise.php

```diff
@@ -28,15 +28,41 @@ public function then(callable $onFulfilled = null, callable $onRejected = null,
             return new static($this->resolver($onFulfilled, $onRejected, $onProgress));
         }
 
-        $this->requiredCancelRequests++;
 
-        return new static($this->resolver($onFulfilled, $onRejected, $onProgress), function () {
-            $this->requiredCancelRequests--;
+        // keep a reference to this promise instance for the static canceller function.
+        // see also parentCancellerFunction() for more details.
+        $parent = $this;
+        ++$parent->requiredCancelRequests;
 
-            if ($this->requiredCancelRequests <= 0) {
-                $this->cancel();
+        return new static($this->resolver($onFulfilled, $onRejected, $onProgress), self::parentCancellerFunction($parent));    }
```

Comment:
CS: missing linebreak before the closing `}` of the method

Reply:
Thanks for spotting, fixed! :shipit: